### PR TITLE
Revived Joint Limits Interfaces with new State and Command Interfaces

### PR DESCRIPTION
**Pull request by guru-florida**
Using new hardware_interface::StateInterface and CommandInterface instead of the old JointHandle class. Mostly a search and replace process at first, but CommandInterface has one surprise where the copy constructor is deleted in favor of always moving. So for command arguments the "const CommandInterface&" becomes "CommandInterface &&" and std::move() is used in implementation. 

The switch to move semantics also invalidates all the tests though since the tests would create a command handle in the test fixture which cannot become a && argument. So I switched the fixture to use inline command_handle() methods to retrieve a command handle which can then be directly passed as an argument to the limit constructors.

### CODE REVIEW

**Comment by guru-florida** [comment]
The joint limit interfaces is just a single include file, not even a src implementation. I could move that file and the test file into the hardware_interface module, I agree it probably belongs in there.

@Karsten1987 mentioned in PR #236 :
> I think that the joint limit interface can safely be ignored for now. I believe that we actually don't need a dedicated package for this. The functionality can easily be incorporated within either the hardware_interface or into a default component package. Interface seems to be wrong in this case then.

**Comment by guru-florida** [comment]
@bmagyar (As per WG meeting, regarding refactor into hardware_interface):
The joint_limits_interface module has a dependency on "urdf" package since it also parses limits out of URDF files. hardware_interface does not (looks like it uses tinyxml2). _urdf_model/joint.h_ and _urdf/urdfdom_compatibility.h_ are indeed included from the _joint_limits_urdf_ header. Are you ok with hardware_interface including the urdf package? No other dependency issues.

**Comment by guru-florida** [comment]
@bmagyar 
The tests are failing due to launch_testing_ament_cmake not being available in the build environment. Possibly the launch repo needs to be added to ci's  vcs-repo-file-url? This package is required by the joint_limits rosparams py test. It's only a BUILD_TESTING dependency.

**Comment by guru-florida** [comment]
Other than test dep issue above, migration of joint_limits_interface into hardware_interface is complete. No significant changes in the API. The joint_limits include files, though moved into hardware_interface, are still in "joint_limits_interface/" include directory so **no code changes for users are required other than dump the joint_limits_interface package ref**.

As for tests, (1) they were converted to gmock instead of gtest to match other hardware_interface tests. (2) The rclcpp package dep had to be included for BUILD_TESTING for a unit tests that actually creates a rclcpp::Node for testing parameters. (3) The urdf package was added to main deps since joint_limits includes urdf parsing.

**Comment by destogl** [comment]
> The joint_limits include files, though moved into hardware_interface, are still in "joint_limits_interface/" include directory
@guru-florida I would propose to add it into hardware_interface/joint_limits_interface/
The main question is how this interface should be used?

@bmagyar we have to rethink this architecture since we do not have "joints" anymore. Probably is this something for ResourceManager

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:20]: Use standardized interface types

```suggestion
#include <hardware_interface/handle.hpp>
#include <hardware_interface/types/hardware_interface_type_values.hpp>
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:49]: use standardized names always when possible

```suggestion
  : jposh_(hardware_interface::HW_IF_POSITION),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:50]:
```suggestion
    jvelh_(hardware_interface::HW_IF_VELOCITY),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:53]: Why not also:

```suggestion
    prev_vel_(std::numeric_limits<double>::quiet_NaN())
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:61]:
```suggestion
    jvelh_(hardware_interface::HW_IF_VELOCITY),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:62]: If you do this here, will this interface be still accessible by controllers and hardware? Do you have a proposal how should `joint_limit_interface` be integrated into the current framework?

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:71]: We should probably use different names to correspond to the new nomenclature:
```suggestion
    const hardware_interface::StateInterface & joint_pos_interface,
    const hardware_interface::StateInterface & joint_vel_interface,
    hardware_interface::CommandInterface && joint_cmd_interface,
```

This should be probably done in a follow-up PR.

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:78]: same as above

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:367]:
```suggestion
      hardware_interface::StateInterface(hardware_interface::HW_IF_VELOCITY),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:440]:
```suggestion
      hardware_interface::StateInterface(hardware_interface::HW_IF_VELOCITY),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:528]:
```suggestion
  : JointLimitHandle(hardware_interface::StateInterface(hardware_interface::HW_IF_POSITION),
```

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [inline hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:529]:
```suggestion
      hardware_interface::StateInterface(hardware_interface::HW_IF_VELOCITY), std::move(jcmdh), limits)
```

File: hardware_interface/test/joint_limits/joint_limits_interface_test.cpp
Comment (destogl) [inline hardware_interface/test/joint_limits/joint_limits_interface_test.cpp:108]:
```suggestion
    return hardware_interface::CommandInterface(name, hardware_interface::HW_IF_POSITION, &cmd);
  }

  inline hardware_interface::StateInterface position_handle()
  {
    return hardware_interface::StateInterface(name, hardware_interface::HW_IF_POSITION, &pos);
  }

  inline hardware_interface::StateInterface velocity_handle()
  {
    return hardware_interface::StateInterface(name, hardware_interface::HW_IF_VELOCITY, &vel);
  }

  inline hardware_interface::StateInterface effort_handle()
  {
    return hardware_interface::StateInterface(name, hardware_interface::HW_IF_EFFORT, &eff);
```

File: hardware_interface/test/joint_limits/joint_limits_interface_test.cpp
Comment (destogl) [inline hardware_interface/test/joint_limits/joint_limits_interface_test.cpp:529]:
```suggestion
    return hardware_interface::CommandInterface(name2, hardware_interface::HW_IF_POSITION, &cmd2);
  }

  inline hardware_interface::StateInterface position2_handle()
  {
    return hardware_interface::StateInterface(name2, hardware_interface::HW_IF_POSITION, &pos2);
  }

  inline hardware_interface::StateInterface velocity2_handle()
  {
    return hardware_interface::StateInterface(name2, hardware_interface::HW_IF_VELOCITY, &vel2);
  }

  inline hardware_interface::StateInterface effort2_handle()
  {
    return hardware_interface::StateInterface(name2, hardware_interface::HW_IF_EFFORT, &eff2);
```

File: hardware_interface/test/joint_limits/joint_limits_rosparam_test.cpp
Comment (destogl) [inline hardware_interface/test/joint_limits/joint_limits_rosparam_test.cpp:32]: Why new line?

**Review by destogl — CHANGES_REQUESTED** [review]
Good job @guru-florida 

I added many comments because it is unclear to me how this would work with the new "component-base" architecture.
Before merging there should be test using `test_single_joint_actuator` and `test_two_joint_system` testing the integration and also an example in the `ros2_control_demos` repository, where we test the whole framework.

Other suggestions/questions:
* Maybe to move `JointLimits-README.md` into corresponding include folder`
* @bmagyar  can we remove `author` tag from joint_limits.hpp since we don't have it in other files...
* Why is there file `joint_limits_rosparam.hpp`, should we allow to set the limits using URDF and parameter server? I find this rather ambiguous...

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (guru-florida) [reply hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:51]: position_command was the name for (queue music) position command interface handles in previous code. So if you were to register a position state and a position command interface it would be "position" and "position_command". Has this changed in components?

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (guru-florida) [reply hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:62]: I assume you meant jpos_h above as well. I wasnt aware of the nomenclature change there, I can easily do that with the refactoring tool.

Comment (guru-florida) [reply hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:62]: I am not fond of this new change where command handles "can only be moved" not copied. I get it, there should only be one owner setting the command, but it makes for some more complex code. So this std::move is a requirement of the explicit deletion of the copy operator in ReadWriteHandle.

File: hardware_interface/test/joint_limits/joint_limits_rosparam_test.cpp
Comment (guru-florida) [reply hardware_interface/test/joint_limits/joint_limits_rosparam_test.cpp:32]: It is using method chaining. Notice no semicolon on the end of line 32....there is another .method() call below it. Read better to me to have the multiple .method() chain calls indented to match and it passed clint/uncrustify.

File: hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp
Comment (destogl) [reply hardware_interface/include/joint_limits_interface/joint_limits_interface.hpp:52]: AFAIK, yes, so we would use "position" only now. I have to check this because `joint_limit_interface` works on the opposite since from controllers.

**Comment by bmagyar** [comment]
@guru-florida did you have time to integrate this with hardware_interface instead? I think this could be closed at this point as we agreed not to have this package separately

**Comment by guru-florida** [comment]
I am updating my own code to use RM/CM now. Let me revisit this again before closing. I am still catching up to the current way of things.

> @guru-florida did you have time to integrate this with hardware_interface instead? I think this could be closed at this point as we agreed not to have this package separately

**Comment by destogl** [comment]
@guru-florida can you please enable "Allow edits and access to secrets by maintainers" box (on the bottom of the right-options of the PR). I would like to push changes directly to your branch.

**Comment by guru-florida** [comment]
I have "Allow edits by maintainers" checked already. There is no "Allow edits and access to secrets...". Apparently that is only when there are GitHub Actions workflows. I'm not sure. It looks like you did get changed pushed here though, I see your HW interface constant changes.

> @guru-florida can you please enable "Allow edits and access to secrets by maintainers" box (on the bottom of the right-options of the PR). I would like to push changes directly to your branch.

**Comment by destogl** [comment]
#441 is the followup

**Pull request closed without merging**
